src/main/java/com/diffblue/demo/ecommerce/controllers/CheckoutController.java

```diff
@@ -54,11 +55,18 @@ public CheckoutController(OrderRepository orderRepo, AddressRepository addressRe
   @RequestMapping("/checkout")
   public String viewCheckoutPage(Map<String, Object> model, HttpSession session,
                                  CustomerForm customerForm) {
-    if (session.getAttribute("shoppingCart") != null) {
+    try {
+      if (session.getAttribute("shoppingCart") == null) {
+        throw new Exception("Shopping Cart does not exist");
```

Comment:
Shall we use custom exceptions for this? Eg `InvalidCartException`

Reply:
As of yesterday, @diffbluejyu is working on an error page for 404s & other exceptions, so yes, i'd like to change that to integrate it with this, when its done. Rather than simply redirecting back to the checkout.